doc/man-source/irexec.1
.TH IREXEC "1" "Last change: Oct 2015" "irexec @version@" "User Commands"
.SH NAME
.P
\fBirexe\fRc - run programs with one button press
.SH SYNOPSIS
.B irexec
[\fIoptions\fR] [\fIconfig_file\fR]
.SH DESCRIPTION
.P
\fBirexec\fR executes commands on an IR signal decoded
by lircd, the LIRC daemon. It uses an lircrc config file where
the \fI config = \fR entries are executed. E. g., given the
following config file snippet
.nf
    begin
        prog   = irexec
        button = KEY_RED
        config = echo "KEY_RED"
    end
.fi
.P
\fBirexec\fR will echo \fIKEY_RED\fR on the terminal when the corresponding
button is pushed on a remote. The command is an arbitrary shell command
executed asynchronously \- \fBirexec\fR does not wait for it to complete.
.SH ARGUMENTS
.TP 4
.B config_file
lircrc configuration file. \fBirexec\fR only uses entries with
\fIprog = irexec\fR. The path defaults to \fI~/.config/lircrc\fR.
.SH OPTIONS
.TP 4
.B -h, --help
Display usage summary
.TP 4
.B -v, --version
Display version
.TP 4
.B  -d, --daemon
Make irexec fork to background. In this case a config file should be given
on the command line as irexec won't be able to find any home directory.
.TP 4
\fB-D, --loglevel\fR [\fIlevel\fR]
Determine the amount of logging information. [\fIlevel\fR] can be a symbolic
syslog level: \fIerror\fR, \fIwarning\fR, \fIinfo\fR, \fInotice\fR or
\fIdebug\fR. lirc also defines three additional levels \fItrace\fR,
\fItrace\fR and \fItrace2\fR which gives even more messages (\fItrace2\fR
bringing the most). However, in the log these messages are marked as
\fIdebug\fR. By default, no logging is done.
.TP 4
\fB-n, --name\fR <\fIname\fR>
Use this program name instead of the default \fIirexec\fR as identifier in
the lircd.conf file.
.SH ENVIRONMENT
.TP 4
.B LIRC_SOCKET_PATH
Path to the lircd socket irexec reads from, defaults to @varrundir@/lirc/lircd.
.SH FILES
.TP 4
.B ~/.config/lircrc
Default config file
.TP 4
.B @etcdir@/lirc/lircrc
Config file used by the systemd irexec service.
.TP 4
.B ~/.cache/irexec.log
Debug log. Setting the XDG_CACHE_HOME environment variable relocates this
file to $XDG_CACHE_HOME/irexec.log
.SH NOTES
.P
For versions up to 0.9.1 \fBirexec\fR used to  wait until the executed program
terminated. Old configuration files thus often includes a  '&' appended
to the command string to avoid being stuck in the command. This is not
required in 0.9.2+ which cannot wait for command completion.
.P
irexec should run as a service. The contrib directory contains a .desktop
file which could be dropped in ~/.local/autostart. Doing so creates a
service which can be handled by regular desktop tools such
as \fIgnome\-tweak\-tool\fR on all major desktops. If running restricted
commands such as powering off the machine is required, sudo(8) can be used
to allow regular users to run such commands.
.P
An alternative is to use the irexec.service systemd service
which runs as root (and can run in parallel with the desktop service).
Running as root has severe security implications.
See the Configuration Guide in the html documentation.
.SH "SEE ALSO"
.TP 4
.B lircrc(5)
The lircrc configuration file.


NEWS
0.9.4-devel
  * Issues: https://sourceforge.net/p/lirc/tickets/milestone/0.9.4

  * The caraca driver has been retired (still available in attic/).
  * The irman driver has been moved out of tree. From 0.5.1 the
    libirman package contains the driver. See drivers/irman/README
  * The yard2 driver has been moved out of tree and is now maintained
    by the yard2 project. See driver/yard2/README.
  * New upstream debian packaging with which Debian users can create
    .deb packages. Ubuntu packages are also available in PPAs. See
     debian/INSTALL (#137).
  * New driver girs for the GIRS protocol, gives LIRC Arduino support.
    See new info in driver documentation.
  * Support for non-modulated signals added (#132).
  * As shipped, lircd now uses the devinput driver and the new 'auto'
    device, should work for all remotes supported by kernel (#106).
  * Enhanced lircd-setup tool (#123, partly)
    - Integrated mode2 and irw tests of current configuration.
    - Documentation links.
    - More consistent data based on new database.
    - Multiple bugfixes.
  * Several patches to make mode2 work with all drivers(#123, partly)
  * Configuration database overhaul (#123, partly)
    - Database lives as yaml files in configs/.
    - Data is collected from drivers, remotes and static files.
    - Better ways for drivers to enumerate devices.
    - There is an unstable python database module used by
      all tools.
  * devinput.lircd.conf is generated at build time to match host kernel,
    new script lirc-make-devinput shipped to re-generate it (#153)
  * A new script lircd-setup runs as a one-shot service before lircd,
    performing setup commands as root (#121).
  * Several new examples in contrib/, notably lircrc templates for
    many popular LIRC applications and improved udev rules.
  * The default driver does not require root perms if the protocol
    is correct (#133).
  * Add an "auto" device to devinput driver for single-device
    usecase (#134).
  * irrecord has been taught to handle remotes without gap (#98).
  * CLI options for mode2, xmode2 and irrecord has been streamlined.
    All tools falls back to lircd defaults where it makes sense.

  * The driver API has been expanded with functions to enumerate
    devices. See configs/README and lib/driver.h.
  * New logging API with hooks for enhanced messages and filtering
    (also #129).
  * The out-of-tree build interface is enhanced with hooks for external
    drivers to rebuild documentation and add a config file.

  * irrecord fixes:
    - Buggy handling of not decoded buttons in irrecord (#146).
    - Repeated buttons problems (#149)
    - Endless loop when key is not recognized (#146)
    - Endless loop on idendifying toggle mark (#139)
  * Remotes become too sensitive in 0.9.3 (#148)
  * Fix --plugindir not being respected (#144).
  * Fix crash for non-existing device (#145).
  * Fix gid and supplementary groups in CLI apps dropping root.
  * Fix a crash in irrecord when invoked without options.
  * Don't require write access if not needed in default.c (#133).
  * Clean up docs, remove circular links (#136). Documentation is
    now only in lirc.org/html.

0.9.3  4/09/15
  * Digest of 0.9.3pre1..0.9.3pre3

  * New build dependency: xsltproc (sometimes coming with libxml).
  * The irexec semantics have been changed so that the invoked
    programs run disconnected from the main irexec process.
    See irexec(1) (#92).
  * irrecord has got some love:
     + It handles multiple definitions of same symbol OK (#78)
     + Warns about single-key config files (#77)
     + Checks for ambient light disturbances (#76)
     + New option allows updating of existing template file.
     + Template handling bug fixed
     + -l --list-namespace option parsing bug fixed (#101).
     + Minor fixes to irrecord-ui including -k option parsing fixed.
  * mode2 has been updated
     +  handle also LIRCCODE devices (#94),
     +  Drop root privileges (#91)
     +  Add  new scope-like display
     +  Handle the--driver-options switch
  * xmode2 has been updated to handle udp driver, the new
    --driver-options switch + some general overhaul.
  * irrecord has been split into a library and a cli interface
    (#36). The library interface is unstable.

  * The select() calls in lircd, lircrcd and library routines have
    been replaced with poll() to avoid crashes when file descriptors
    becomes > 1024 (#109).
  * Set supplementary groups using effective-user (RH bz 1253907).
  * Work around kernel bug in /dev/lirc0 output (#119).
  * lircmd: Several option parsing bugs fixed, one of which
    blocked the systemd service (#104, RH bz 1234155).
  * lirc-lsremotes now displays logs after parsing files, making
    more parsing warnings visible.
  * lirc-config-tool is fixed to comply with guidelines (#105).
  * Bug in irexec when using modes fixed (#102).
  * Create log directory if required (#124).
  * lirc-setup: Fixed name clash on recent python versions.

  * irtoy: New support for debug LEDs. New documentation
    file, support in lirc-setup and driver table doc added.
  * The udp driver is patched to allow adjustable clockrates,
    see new udp_driver.html driver doc.
  * The mplay driver has been updated with patches which has
    been available "out there".

  * The library routines for parsing lircd.conf now warns for
    duplicate keys and values (#107).
  * Enhanced pkg-config support for drivers compiled out-of-tree
    (see driver-api in the manual).
  * Hardcoded CFLAGS in Makefiles fixed (#124).
  * The compiler has been switched to c++, and filenames are
    updated to *.cpp (#79)
  * Build: now uses parallel make by default.
  * Overall code formatting is improved with more consistent
    layout and updated git hook. See CONTRIBUTE.md

  * Unlinked and hard to find driver docs and manpages have
    been linked.
  * The documentation is heavily revised. Old obsolete text is

tools/irexec.cpp
/****************************************************************************
** irexec.c ****************************************************************
****************************************************************************
*
* irexec  - execute programs according to the pressed remote control buttons
*
* Copyright (C) 1998 Trent Piepho <xyzzy@u.washington.edu>
* Copyright (C) 1998 Christoph Bartelmus <lirc@bartelmus.de>
*
*/

#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <errno.h>
#include <unistd.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/wait.h>

#include "lirc_client.h"
#include "lirc_log.h"

static const logchannel_t logchannel = LOG_APP;

static const char* const USAGE =
	"Usage: irexec [options] [lircrc config_file]\n"
	"\t-d --daemon\t\tRun in background\n"
	"\t-D --loglevel=level\t'error', 'info', 'notice',... or 0..10\n"
	"\t-n --name=progname\tUse this program name for lircrc matching\n"
	"\t-h --help\t\tDisplay usage summary\n"
	"\t-v --version\t\tDisplay version\n";

static const struct option options[] = {
	{ "help",     no_argument,	 NULL, 'h' },
	{ "version",  no_argument,	 NULL, 'v' },
	{ "daemon",   no_argument,	 NULL, 'd' },
	{ "name",     required_argument, NULL, 'n' },
	{ "loglevel", required_argument, NULL, 'D' },
	{ 0,          0,		 0,    0   }
};

static int opt_daemonize	= 0;
static loglevel_t opt_loglevel	= LIRC_NOLOG;
static const char* opt_progname	= "irexec";

static char path[256] = {0};


/** Run shell command line in isolated process using double fork(). */
static void run_command(const char* cmd)
{
	pid_t pid1;
	pid_t pid2;

	pid1 = fork();
	if (pid1 < 0) {
		log_perror_err("Cannot fork");
		perror("Cannot fork()");
		exit(EXIT_FAILURE);
	}
	if (pid1 == 0) {
		pid2 = fork();
		if (pid2 < 0) {
			log_perror_err("Cannot do secondary fork()");
			exit(EXIT_FAILURE);
		}
		if (pid2 == 0) {
			log_debug("Execing command \"%s\"", cmd);
			char* const vp[] = {strdup(SH_PATH),
			      strdup("-c"),
			      strdup(cmd),
			      NULL
			};
			execvp(SH_PATH, vp);
			/* not reached */
			log_perror_err("execvp failed");
			fputs("execvp failed\n", stderr);
		} else {
			waitpid(pid2, NULL, WNOHANG);
			exit(0);
		}
	} else {
		waitpid(pid1, NULL, 0);
	}
}


/** Get buttonclick messages from lircd socket and process them. */
static void process_input(struct lirc_config* config)
{
	char* code;
	char* c;
	int r;

	while (lirc_nextcode(&code) == 0) {
		if (code == NULL)
			continue;
		r = lirc_code2char(config, code, &c);
		while (r == 0 && c != NULL) {
			run_command(c);
			r = lirc_code2char(config, code, &c);
		}
		free(code);
		if (r == -1)
			break;
	}
}


int irexec(const char* configfile)
{
	struct lirc_config* config;

	if (opt_daemonize) {
		if (daemon(0, 0) == -1) {
			perror("Can't daemonize");
			return EXIT_FAILURE;
		}
	}
	if (lirc_init(opt_progname, opt_daemonize ? 0 : 1) == -1)
		return EXIT_FAILURE;

	if (lirc_readconfig(configfile, &config, NULL) != 0) {
		fputs("Cannot parse config file\n", stderr);
		return EXIT_FAILURE;
	}
	lirc_log_get_clientlog("irexec", path, sizeof(path));
	unlink(path);
	lirc_log_set_file(path);
	lirc_log_open("irexec", 1, opt_loglevel);

	process_input(config);
	lirc_deinit();

	lirc_freeconfig(config);
	return EXIT_SUCCESS;
}


int main(int argc, char* argv[])
{
	int c;

	while ((c = getopt_long(argc, argv, "D:hvdn:", options, NULL)) != -1) {
		switch (c) {
		case 'h':
			puts(USAGE);
			return EXIT_SUCCESS;
		case 'v':
			puts("irexec " VERSION);
			return EXIT_SUCCESS;
		case 'd':
			opt_daemonize = 1;
			break;
		case 'n':
			opt_progname = optarg;
			break;
		case 'D':
			opt_loglevel = string2loglevel(optarg);
			break;
		default:
			fputs(USAGE, stderr);
			return EXIT_FAILURE;
		}
	}
	if (optind < argc - 1) {
		fputs("Too many arguments\n", stderr);
		return EXIT_FAILURE;
	}
	if (opt_loglevel == LIRC_BADLEVEL) {
		fprintf(stderr, "Bad debug level: %s\n", optarg);
		return EXIT_FAILURE;
	}
	return irexec(optind != argc ? argv[optind] : NULL);
}
